File: 2024/day_8/part2.ts
// https://adventofcode.com/2024/day/1

import { readInput } from "../../utils/utils";

const input = await readInput();
const lines = input.split("\n").filter(Boolean);
const spaces = lines.map(line => line.split(""));
type Point = { x: number; y: number };
const antenae = spaces
  .flatMap((line, y) =>
    line
      .map((char, x) => (![".", "#"].includes(char) ? { x, y, char } : null))
      .filter(x => x !== null)
  )
  .reduce(
    (acc, cur) => ((acc[cur.char] ??= []).push(cur), acc),
    {} as { [char: string]: Point[] }
  );

function isAntinode(x: number, y: number): boolean {
  for (const points of Object.values(antenae)) {
    if (points.some(p => p.x === x && p.y === y)) return true;
    for (let i = 0; i < points.length; i++) {
      for (let j = i + 1; j < points.length; j++) {
        const p1 = points[i]!,
          p2 = points[j]!;

        const angle1 = Math.atan2(y - p1.y, x - p1.x);
        const angle2 = Math.atan2(y - p2.y, x - p2.x);
        if (angle1 !== angle2) continue;
        return true;
      }
    }
  }
  return false;
}

// Note, points can't be equal
const antinodes = new Set<string>();
for (let y = 0; y < spaces.length; y++) {
  for (let x = 0; x < spaces[y]!.length; x++) {
    if (isAntinode(x, y)) antinodes.add(`${x},${y}`);
  }
}
console.log(antinodes.size);

File: 2024/day_8/Part2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.stream.IntStream;

public class Part2 {
  private static Scanner readInput(String[] args) throws FileNotFoundException {
    if (args.length == 0) return new Scanner(System.in);
    return new Scanner(new File(args[0]));
  }

  public static boolean isAntinode(int x, int y, Map<Character, List<Point>> antennae) {
    // System.out.println("Testing " + x + "," + y);
    for (List<Point> points : antennae.values()) {
      if (points.contains(new Point(x, y))) return true;
      // the angle to two points is the same , we are in the antennae
      for (int i = 0; i < points.size(); i++) {
        for (int j = i + 1; j < points.size(); j++) {
          // System.out.printf("Testing(%d,%d): %s, %s\n", i, j, points.get(i), points.get(j));
          Point p1 = points.get(i), p2 = points.get(j);

          // Check both in same direction
          double angleToP1 = Math.atan2(y - p1.y(), x - p1.x());
          double angleToP2 = Math.atan2(y - p2.y(), x - p2.x());
          // System.out.printf("Angle: %f, %f\n", angleToP1, angleToP2);
          if (angleToP1 != angleToP2) continue;
          return true;
        }
      }
    }
    return false;
  }

  public static void main(String[] args) throws FileNotFoundException {
    // Grid index is [y][x]
    List<List<Character>> grid = new ArrayList<>();

    try (Scanner scan = readInput(args)) {
      while (scan.hasNextLine()) {
        String line = scan.nextLine();
        if (line.isEmpty()) continue;
        line = line.replaceAll("#", "."); // This is an artifact of the input format
        grid.add(line.chars().mapToObj(x -> (char) x).toList());
      }
    }
    // Cache the points of the antennae
    Map<Character, List<Point>> antennae = new HashMap<>();
    for (int y = 0; y < grid.size(); y++) {
      for (int x = 0; x < grid.get(y).size(); x++) {
        char c = grid.get(y).get(x);
        if (c == '.') continue;
        antennae.computeIfAbsent(c, k -> new ArrayList<>()).add(new Point(x, y));
      }
    }

    long count = IntStream.range(0, grid.size()).boxed().flatMap(y -> IntStream.range(0, grid.get(y).size())
        .filter(x -> isAntinode(x, y, antennae)).mapToObj(x -> new Point(x, y))).distinct().count();
    System.out.println(count);

  }
}

record Point(int x, int y) {}
